compiler/rustc_transmute/src/maybe_transmutable/mod.rs

```diff
@@ -153,230 +148,136 @@ where
-                                    let start = cmp::max(*sv.start(), *dst_validity.start());
-                                    let end = cmp::min(*sv.end(), *dst_validity.end());
-
-                                    // We add 1 here to account for the fact
-                                    // that `end` is an inclusive bound.
-                                    accum.sum += 1 + usize::from(end.saturating_sub(start));
-                                } else {
-                                    accum.found_uninit = true;
-                                }
-                                accum
-                            };
-
-                            let answers = self
-                                .dst
-                                .states_from(dst_state, src_validity.clone())
-                                .map(move |(dst_validity, dst_state_prime)| {
-                                    let mut c = c.borrow_mut();
-                                    accum1.set(update_accum(accum1.get(), dst_validity));
-                                    let answer =
-                                        self.answer_memo(&mut *c, src_state_prime, dst_state_prime);
-                                    answer
+            let bytes_answer = src_quantifier.apply(
+                union(self.src.bytes_from(src_state), self.dst.bytes_from(dst_state)).map(
+                    |(_range, (src_state_prime, dst_state_prime))| match (
+                        src_state_prime,
+                        dst_state_prime,
+                    ) {
+                        // No matching transitions in `src`. Ignore.
+                        (None, _) => src_quantifier.empty(),
+                        // No matching transitions in `dst`. Fail.
+                        (Some(_), None) => Answer::No(Reason::DstIsBitIncompatible),
+                        // Transitions in both `src` and `dst`. Continue with successor states.
+                        (Some(src_state_prime), Some(dst_state_prime)) => {
+                            self.answer_memo(cache, src_state_prime, dst_state_prime)
+                        }
+                    },
+                ),
+            );
+
+            // The below early returns reflect how this code would behave:
+            //   if self.assume.validity {
+            //       or(bytes_answer, refs_answer)
+            //   } else {
+            //       and(bytes_answer, refs_answer)
+            //   }
+            // ...if `refs_answer` was computed lazily. The below early
+            // returns can be deleted without impacting the correctness of
+            // the algorithm; only its performance.
+            debug!(?bytes_answer);
+            match bytes_answer {
+                Answer::No(_) if !self.assume.validity => return bytes_answer,
+                Answer::Yes if self.assume.validity => return bytes_answer,
+                _ => {}
+            };
```

Comment:
Does this behave correctly when there are no byte transitions in the source type? It would need to reach the `_` arm. In particular, `Answer::Yes` would be invalid since `assume.validity` is supposed to only return yes when there is at least one source value which can be transmuted.

IIUC, this code path is unreachable for uninhabited types (we handle those elsewhere), so if you get to this point in the code and find no byte transitions, then there must be at least one ref transition. That implies that transmutability should solely be a function of the ref transitions, and so bailing early with `Answer::Yes` (when `self.assume.validity`) is invalid.

Reply:
Can we add a test for this, both for `self.assume.validity` and `!self.assume.validity`?

I realize now why this is the case - it bottoms out at `Quantifier::empty`. But IMO it's confusing to bottom out that way because this `match` is effectively written in a way that relies on the specific implementation of `empty`.

Maybe we could instead modify `Quantifier::apply` to take a default `Answer` that is used if the provided `iter` is empty? IMO that would make this usage here clearer.

Reply:
Okay sounds good. Feel free to follow up with that change in a separate PR.